### PR TITLE
etcdmain, pkg: Support peer and client TLS auth based on SAN fields.

### PR DESCRIPTION
**Pull request by jmillikin-stripe**
Etcd currently supports validating peers based on their TLS certificate's
CN field. The current best practice for creation and validation of TLS
certs is to use the Subject Alternative Name (SAN) fields instead, so that
a certificate might be issued with a unique CN and its logical
identities in the SANs.

This commit extends the peer validation logic to use Go's
`(*"crypto/x509".Certificate).ValidateHostname` function for name
validation, which allows SANs to be used for peer access control.

In addition, it allows name validation to be enabled on clients as well.
This is used when running Etcd behind an authenticating proxy, or as
an internal component in a larger system (like a Kubernetes master).

(See issue https://github.com/etcd-io/etcd/issues/8262 and PR https://github.com/etcd-io/etcd/pull/8616 for additional context)

### CODE REVIEW

**Comment by jingyih** [comment]
cc @wenjiaswe

File: Documentation/op-guide/configuration.md [hunk 1]
```diff
@@ -270,6 +270,11 @@ The security flags help to [build a secure etcd cluster][security].
 + default: ""
 + env variable: ETCD_CLIENT_CRL_FILE
 
+### --client-cert-allowed-name
```
Comment (xiang90) [inline Documentation/op-guide/configuration.md]: allowed-san?

**Comment by xiang90** [comment]
@jmillikin-stripe 

sorry for the long delay. this pr looks good to me. but can we change allowed-name to allowed-san or something more specific?

File: Documentation/op-guide/configuration.md (hunk 1 above)
Comment (jmillikin-stripe) [reply Documentation/op-guide/configuration.md]: When checking whether a given hostname is valid for a TLS certificate, it's typical to not mandate which particular field the name comes from. The [implementation of `(*x509.Certificate).VerifyHostname()`](https://github.com/golang/go/blob/go1.12.6/src/crypto/x509/verify.go#L994-L1028) uses SAN if any are set, and falls back to CN.

Comment (xiang90) [reply Documentation/op-guide/configuration.md]: OK. The implementation checks CN first, then checks parsed SAN, right? Can we make this clear?

Comment (xiang90) [reply Documentation/op-guide/configuration.md]: It is hard to infer this behavior from the flag and current help message.

Comment (jmillikin-stripe) [reply Documentation/op-guide/configuration.md]: I believe the verification logic is something like this:
```
func checkName(name string) bool {
  if len(SAN) == 0 {
    return validHostname(CN) && matches(CN, name)
  }
  for _, sanEntry := range SAN {
    if matches(sanEntry, name) {
      return true
    }
  }
  return false
}
```

The full behavior is described by [RFC 6125 section 6.4](https://tools.ietf.org/html/rfc6125#section-6.4), which is too long to describe here.

**Comment by wenjiaswe** [comment]
@xiang90 If I understood right, x509's logic is, if a SSL Certificate has a Subject Alternative Name (SAN) field, then SSL clients are supposed to ignore the Common Name value and seek a match in the SAN list. As in RFC 6125, Appendix B.2.:
```
   If a subjectAltName extension of type dNSName is present, that MUST
   be used as the identity.  Otherwise, the (most specific) Common Name
   field in the Subject field of the certificate MUST be used.  Although
   the use of the Common Name is existing practice, it is deprecated and
   Certification Authorities are encouraged to use the dNSName instead.
```
 I think in this PR, only one of `--peer-cert-allowed-cn` and `--peer-cert-allowed-name` can be set at a time. If `--peer-cert-allowed-cn` is set, then CommonName is used and it follows the old behavior, if new `--peer-cert-allowed-name` is used, then it follows "current best practice" to check SAN and ignore CN if SAN is set, otherwise check CN. @jmillikin-stripe did I understand correctly?

If that's the case, maybe we could use --peer-cert-allowed-**hostname** instead of --peer-cert-allowed-**name** to match (*x509.Certificate).VerifyHostname() function name?

**Comment by jmillikin-stripe** [comment]
@wenjiaswe Yep, that all matches my understanding. I'd be fine with renaming the new flags to `-hostname`.

File: pkg/transport/listener.go [hunk 2]
```diff
@@ -362,6 +364,7 @@ func (info TLSInfo) ServerConfig() (*tls.Config, error) {
 	}
 
 	cs := info.cafiles()
+	fmt.Fprintf(os.Stderr, "== CAFILES: %v\n", cs)
```
Comment (xiang90) [inline pkg/transport/listener.go]: remove ?

File: pkg/transport/listener.go [hunk 3]
```diff
@@ -264,27 +264,29 @@ func (info TLSInfo) baseConfig() (*tls.Config, error) {
...
-	if info.AllowedName != "" {
+	if info.AllowedHostname != "" {
 		verifyCertificate = func(cert *x509.Certificate) bool {
-			return cert.VerifyHostname(info.AllowedName) == nil
+			return cert.VerifyHostname(info.AllowedHostname) == nil
 		}
 	}
 	if verifyCertificate != nil {
 		cfg.VerifyPeerCertificate = func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+			fmt.Fprintf(os.Stderr, "== VERIFY PEER CERT (%d verified chains)\n", len(verifiedChains))
```
Comment (xiang90) [inline pkg/transport/listener.go]: remove?

File: pkg/transport/listener_tls.go [hunk 4]
```diff
@@ -98,6 +99,7 @@ func checkSAN(ctx context.Context, tlsConn *tls.Conn) error {
 	st := tlsConn.ConnectionState()
 	if certs := st.PeerCertificates; len(certs) > 0 {
 		addr := tlsConn.RemoteAddr().String()
+		fmt.Fprintf(os.Stderr, "checkSAN(addr=%q, certs=%v)\n", addr, certs)
```
Comment (xiang90) [inline pkg/transport/listener_tls.go]: remove?

File: pkg/transport/listener.go (hunk 2 above)
Comment (jmillikin-stripe) [reply pkg/transport/listener.go]: Done.

File: pkg/transport/listener.go (hunk 3 above)
Comment (jmillikin-stripe) [reply pkg/transport/listener.go]: Done.

File: pkg/transport/listener_tls.go (hunk 4 above)
Comment (jmillikin-stripe) [reply pkg/transport/listener_tls.go]: Done.

**Comment by jmillikin-stripe** [comment]
Stray debugging prints removed, sorry about that.

**Comment by jmillikin-stripe** [comment]
The test failure in `["TARGET=linux-amd64-integration-4-cpu"]` looks flaky -- can an etcd maintainer re-try that test?

**Comment by jingyih** [comment]
> The test failure in `["TARGET=linux-amd64-integration-4-cpu"]` looks flaky -- can an etcd maintainer re-try that test?

=== RUN   TestBalancerUnderBlackholeKeepAliveWatch
--- FAIL: TestBalancerUnderBlackholeKeepAliveWatch (5.73s)
    black_hole_test.go:83: took too long to receive watch events

This is a known flaky test, do not worry about it.

**Comment by jingyih** [comment]
The semaphoreci tests timed out after 20m. Can you rebase to master? It was adjusted to 30m.

**Comment by xiang90** [comment]
LGTM if all failures are known.

**Comment by jmillikin-stripe** [comment]
The semaphore test is consistently failing with a timeout in the new test. Please don't merge until I can debug it.

**Comment by jmillikin-stripe** [comment]
SemaphoreCI is green now. I've discovered an interesting behavior in the Go TLS library: the error message returned from a failed handshake is not deterministic. When both sides of a TLS handshake distrust the other, the error races between `"client certificate authentication failed"` and `"remote error: tls: bad certificate"` depending on which side sends a NACK frame.

The test for CN tends toward the second error, and the new test tends toward the first.

**Comment by jmillikin-stripe** [comment]
Gentle ping

**Comment by jmillikin-stripe** [comment]
Thanks!

**Pull request merged**